### DISCUSSION TITLE
Scale down ScaledJob Jobs even when they have not completed

### DISCUSSION THREAD

**Discussion by blaizmannen** [kedacore/keda#4136] [General]
We converted our Sidekiq(asynchronous job processor for ruby) deployment to a KEDA ScaledJob to tackle the following problem:

- scale down operations of our deployment causes some long running jobs to terminate prematurely. Solving that by making the `terminationGracePeriodSeconds` into a large value (i.e 3600) could work, but this would also block any new helm upgrades/new deployments from going through. 

Hence we decided to use KEDA ScaledJob as a workaround, since jobs are independent of each other and a long running Job will not block a new helm upgrade of an app. 

It works, but since our Sidekiq isn't a Job by nature, it never ever completes/exits and keeps running while waiting for new tasks. This prevents KEDA ScaledJob from scaling the Jobs down. Scaling up works fine.

What we could do is make the ScaledJob Job run a command where it checks if there are no active sidekiq jobs being processed for X consecutive seconds, then terminate the process and complete the Job. But before we do that, we would like to know if there is any native way that KEDA can achieve this, being allow the KEDA ScaledJob to scale down the Jobs even if they have not completed.

**Comment by JCMais** [comment]
did you figure this out? I need to do something similar with BullMQ

**Comment by SpiritZhou** [comment]
What about adding a "timeout" feature in ScaledJob and scale down jobs when it exceed the timeout? @tomkerkhove @JorTurFer @zroubalik

**Reply by zroubalik** [reply]
@SpiritZhou Adding a timetout might be great addition, though I am not sure if it fully solves the issue.

@blaizmannen sorry, I missed this somehow. I suppose you went with the proposed apporach, right?

**Reply by JCMais** [reply]
the timeout was basically what I did in the application layer. Having this as part of Keda would be great yeah.

The timeout would just scale the number of jobs down, but it should not kill the jobs, just signal for them to exit, as they may be processing a task atm

**Reply by salmonsteakk** [reply]
Hey could I check if there are plans for this on the roadmap? I'm thinking of doing the same thing but with solid queue, but from what I understand, scaledJob doesn't seem to include a way to kind of scale the pods back down to 0 and expects the pods to handle the termination logic themselves upon completion. I really like the "independent" nature of scaledJobs which allows us to run long-running jobs, but would also love to have the feature of scaling down too

**Comment by blaizmannen** [comment]
To solve the problem, we ended up not going with KEDA.

Our sidekiq application remained as a kubernetes deployment. We then created a preStopHook script that would send a TSTP signal to the sidekiq application, and wait X amount of seconds OR wait till there are no more sidekiq jobs (whichever first), before the TERM and then ultimately the KILL signals are sent. This is enough to gracefully terminate our sidekiq pods in the event of any scale down operation/deployment update, and allow the jobs to be rescheduled back into the queue if they get interrupted

**Reply by JorTurFer** [reply]
you can do exactly this but scaling out/in based on the queue length xD
I mean, the goal of KEDA is driving the autoscaling, if you need the autoscaling feature and you have solved the pod shutdown, you can include a ScaledObject for scaling in/out (if you need it)
